tutorials/how-to-implement-rag-generativeapis/index.mdx

```diff
@@ -70,288 +77,285 @@ Create a .env file and add the following variables. These will store your API ke
 
-### Load metadata for improved efficiency
-
-By loading the metadata for all objects in your bucket, you can speed up the process significantly. This allows you to quickly check if a document has already been embedded without the need to load the entire document.
+   ```python
+   from langchain_community.document_loaders import S3DirectoryLoader
+   from langchain.text_splitter import RecursiveCharacterTextSplitter
+   ```
 
-```python
-# rag.py
-
-session = boto3.session.Session()
-client_s3 = session.client(service_name='s3', endpoint_url=os.getenv("SCW_BUCKET_ENDPOINT", ""),
-                               aws_access_key_id=os.getenv("SCW_ACCESS_KEY", ""),
-                               aws_secret_access_key=os.getenv("SCW_API_KEY", ""))
-paginator = client_s3.get_paginator('list_objects_v2')
-page_iterator = paginator.paginate(Bucket=os.getenv("SCW_BUCKET_NAME", ""))
-```
+### Iterate through objects
+
+Edit `embed.py` to load all files in your bucket using `S3DirectoryLoader`, split them into chunks of 500 characters using `RecursiveCharacterTextSplitter` and embed them and store them into your PostgreSQL database using `PGVector`.
+
+   ```python
+   text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=0, add_start_index=True, length_function=len, is_separator_regex=False)
+   file_loader = S3DirectoryLoader(
+       bucket=os.getenv("SCW_BUCKET_NAME"),
+       prefix="",
+       endpoint_url=os.getenv("SCW_BUCKET_ENDPOINT"),
+       aws_access_key_id=os.getenv("SCW_ACCESS_KEY"),
+       aws_secret_access_key=os.getenv("SCW_SECRET_KEY")
+   )
+   for file in file_loader.lazy_load():
+       chunks = text_splitter.split_text(file.page_content)
+       embeddings_list = [embeddings.embed_query(chunk) for chunk in chunks]
+       vector_store.add_embeddings(chunks, embeddings_list)
+       print('Vectors successfully added for document',file.metadata['source'])
+   ```
 
-In this code sample, we:
-- Set up a Boto3 session: we initialize a Boto3 session, which is the AWS SDK for Python, fully compatible with Scaleway Object Storage. This session manages configuration, including credentials and settings, that Boto3 uses for API requests.
-- Create an Amazon S3 client: we establish an Amazon client to interact with the Scaleway Object Storage service.
-- Set up pagination for listing objects: we prepare pagination to handle potentially large lists of objects efficiently.
-- Iterate through the bucket: this initiates the pagination process, allowing us to list all objects within the specified Scaleway Object bucket seamlessly.
+   The chunk size of 500 characters is chosen to fit within the context size limit of the embedding model used in this tutorial, but could be raised to up to 4096 characters for `bge-multilingual-gemma2` model (or slightly more as context size is counted in tokens). Keeping chunks small also optimizes performance during inference.
```

Comment:
The embedding model used in this tutorial is now `bge-multilingual-gemma2` perhaps we need to edit this line accordingly, as well as the chunk_size defined above